src/basic/util.c

```diff
@@ -52,32 +52,57 @@ int prot_from_flags(int flags) {
 }
 
 bool in_initrd(void) {
-        struct statfs s;
         int r;
+        const char *e;
 
         if (saved_in_initrd >= 0)
                 return saved_in_initrd;
 
-        /* We make two checks here:
+        /* We have two checks here:
          *
          * 1. the flag file /etc/initrd-release must exist
          * 2. the root file system must be a memory file system
          *
          * The second check is extra paranoia, since misdetecting an
          * initrd can have bad consequences due the initrd
          * emptying when transititioning to the main systemd.
+         *
+         * If env var $SYSTEMD_IN_INITRD is not set or set to "auto",
+         * both checks are used. If it's set to "lenient", only check
+         * 1 is used. If set to a booleen value, then the boolean
+         * value is returned.
          */
 
-        r = getenv_bool_secure("SYSTEMD_IN_INITRD");
-        if (r < 0 && r != -ENXIO)
+        e = secure_getenv("SYSTEMD_IN_INITRD");
+        if (e) {
+                r = parse_boolean(e);
+                if (r >= 0) {
+                        saved_in_initrd = r > 0;
+                        return saved_in_initrd;
+                }
+
+                if (streq(e, "auto"))
+                        goto check_all;
+                else if (streq(e, "lenient"))
+                        goto check_etc;
+
                 log_debug_errno(r, "Failed to parse $SYSTEMD_IN_INITRD, ignoring: %m");
+        }
 
-        if (r >= 0)
-                saved_in_initrd = r > 0;
-        else
-                saved_in_initrd = access("/etc/initrd-release", F_OK) >= 0 &&
-                                  statfs("/", &s) >= 0 &&
-                                  is_temporary_fs(&s);
+check_all:
+        r = path_is_temporary_fs("/");
+        if (r < 0)
+                log_error_errno(r, "Couldn't determine if / is a temporary file system: %m");
+
+        saved_in_initrd = r > 0;
+check_etc:
+        r = access("/etc/initrd-release", F_OK);
+        if (r >= 0) {
+                if (saved_in_initrd == 0)
+                        log_warning("/etc/initrd-release exists, but it's not an initrd.");
```

Comment:
won't this warning be printed in your use case?

Reply:
this one should be downgraded to debug level, too, and i guess can stay then

Reply:
> won't this warning be printed in your use case?

Hi, this will be printed only if SYSTEMD_IN_INITRD is auto or unset, else, svaed_in_initrd here should be -1.

> this one should be downgraded to debug level, too, and i guess can stay then

I'll fix the log level.